### PR TITLE
Extra `HashableScriptData` in `TxBody` for PK inputs and reference inputs with hashed datums.

### PR DESCRIPTION
**Pull request by mmontin**
This PR aims to solve issue #600 by allowing users to provide instances of `HashableScriptData` for hashed datums located:
* in reference inputs
* in regular inputs owned by private keys

For now, these data can only be provided:
* in outputs
* in regular inputs owned by scripts

This contribution intends to be as little invasive as possible, and to respect the current way of handling the creation of the data map provided in the script context. The map is currently built by browsing through the relevant parts of the body content and retrieving the relevant data there. Thus, the idea here is not to create a new field in the body content, but instead to allow user to add additional `HashableScriptData` at two addition locations in the body content.

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Script.hs [hunk 1]
```diff
@@ -722,6 +724,15 @@ data ScriptWitness witctx era where
 
 deriving instance Show (ScriptWitness witctx era)
 
+data PrivateKeyWitness witctx era where
+  PrivateKeyWitness
+    :: KeyDatum witctx
+    -> PrivateKeyWitness witctx era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Script.hs:730]: Can you make this not a GADT? I know there are lots of them floating around but we are actually moving away from using GADTs when not needed.

File: cardano-api/internal/Cardano/Api/Script.hs [hunk 2]
```diff
@@ -773,6 +784,14 @@ deriving instance Eq (ScriptDatum witctx)
 
 deriving instance Show (ScriptDatum witctx)
 
+data KeyDatum witctx where
+  KeyDatumForTxIn :: Maybe HashableScriptData -> KeyDatum WitCtxTxIn
+  NoKeyDatumForStake :: KeyDatum WitCtxStake
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Script.hs:789]: Can you make this not a GADT? I know there are lots of them floating around but we are actually moving away from using GADTs when not needed.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Sorry this is not the cleanest approach to add this functionality. Myself and my team will handle this. Thank you for the attempt!

File: cardano-api/internal/Cardano/Api/Script.hs (hunk 1 above)
Comment (mmontin) [reply cardano-api/internal/Cardano/Api/Script.hs:730]: Yes i can. I basically just mimicked how other types around were defined. But indeed GADTs are not necessary there.

**Comment by mmontin** [comment]
@Jimbo4350 This was just the very beginning and I was in need for feedback but that's fine by me if you handle it.

**Comment by Jimbo4350** [comment]
@mmontin https://github.com/IntersectMBO/cardano-api/pull/640

**Pull request closed without merging**